tuxedo-template-runtime/src/kitties.rs

```diff
@@ -1,126 +1,544 @@
-#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
-pub struct KittiesPiece; // Decodes Value -> KittyData
-impl TuxedoPiece for KittiesPiece {
-    type Data = KittyData; // This is API user Defined
-    const TYPE_ID: TypeId = *b"2222";
-    type Error = ();
-
-    fn validate(&self, transaction: Transaction) -> Result<(), Self::Error> {
-        PreValidator::<Self>::pre_validate(&transaction)?;
-        // TODO: Implement Kitty Logic scenario
-
-        // 1.) If you want to breed Mom cannot have given birth before
-        // 2.) If you want to breed Dad cannot be too tired
-        for input in transaction.inputs.iter() {
-            let parents = PieceExtracter::<Self>::extract(input.output)?;
-            Self::mom_and_dad_ready(&parents)?;
-        }
+
+impl UtxoData for KittyData {
+    const TYPE_ID: [u8; 4] = *b"Kitt";
+}
+
+#[cfg_attr(
+    feature = "std",
+    derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf)
+)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub enum VerifierError {
+    /// Dynamic typing issue.
+    /// This error doesn't discriminate between badly typed inputs and outputs.
+    BadlyTyped,
+    /// Needed when spending for breeding.
+    MinimumSpendAndBreedNotMet,
+    /// Need two parents to breed.
+    TwoParentsDoNotExist,
+    /// Incorrect number of outputs when it comes to breeding.
+    NotEnoughFamilyMembers,
+    /// Mom has recently given birth and isnt ready to breed.
+    MomNotReadyYet,
+    /// Dad cannot breed because he is still too tired.
+    DadTooTired,
+    /// Cannot have two moms when breeding.
+    TwoMomsNotValid,
+    /// Cannot have two dads when breeding.
+    TwoDadsNotValid,
+    /// New Mom after breeding should be in HadBirthRecently state.
+    NewMomIsStillRearinToGo,
+    /// New Dad after breeding should be in Tired state.
+    NewDadIsStillRearinToGo,
+    /// Number of free breedings of new parent is not correct.
+    NewParentFreeBreedingsIncorrect,
+    /// New parents DNA does not match the old one parent has to still be the same kitty.
+    NewParentDnaDoesntMatchOld,
+    /// New parent Breedings has not incremented or is incorrect.
+    NewParentNumberBreedingsIncorrect,
+    /// New child DNA is not correct given the protocol.
+    NewChildDnaIncorrect,
+    /// New child doesnt have the correct number of free breedings.
+    NewChildFreeBreedingsIncorrect,
+    /// New child has non zero breedings which is impossible because it was just born.
+    NewChildHasNonZeroBreedings,
+    /// New child parent info is either in Tired state or HadBirthRecently state which is not possible.
+    NewChildIncorrectParentInfo,
+    /// Too many breedings for this kitty can no longer breed.
+    TooManyBreedingsForKitty,
+    /// Not enough free breedings available for these parents.
+    NotEnoughFreeBreedings,
+}
+
+trait Breed {
```

Comment:
Why is this a trait instead of just methods on the `FreeKittyVerifier` or even just helper functions. Do you imagine there will be another implementation at some point?

Reply:
It is possible I was leaving the possibility open. Also was nice for having some associate types encapsulated